### PR TITLE
bitpay v2 updates

### PR DESCRIPTION
**Pull request by thejoshualewis**
Updated to use BitPay v2 API if conditions match.

### CODE REVIEW

File: lib/offsite_payments/integrations/bit_pay.rb [hunk 1]
```diff
@@ -23,12 +23,19 @@ class Helper < OffsitePayments::Helper
...
           add_field('fullNotifications', true)
           add_field('transactionSpeed', 'high')
         end
 
+        if options[:checkout_token].length >=44
+            self.invoicing_url = 'https://bitpay.com/invoices'
+            #check for characters and revert
+            if options[:checkout_token].include? "0" || options[:checkout_token].include? "O" || options[:checkout_token].include? "I" || options[:checkout_token].include? "l"
+              self.invoicing_url = 'https://bitpay.com/api/invoice'
+            end
```
Comment (pi3r) [inline lib/offsite_payments/integrations/bit_pay.rb]: Can you add a test for it? I'd like to understand what's the actual difference between the two type of tokens.

**Comment by thejoshualewis** [comment]
The older v1 tokens have a max length of 42 characters, the v2 have a max of 44, so what that extra code section does is look at the token that’s saved in the config (I’m assuming since I can’t really test in a local sandbox) and if it matches the v2 standards, it switches the endpoint to the v2 version, otherwise it falls back to the v1

**Comment by pi3r** [comment]
Can you write a test case for it? That will help explain what's happening, also why you're checking all those characters.

File: lib/offsite_payments/integrations/bit_pay.rb [hunk 2]
```diff
@@ -23,12 +23,19 @@ class Helper < OffsitePayments::Helper
         def initialize(order_id, account, options)
           super
           @account = account
-
           add_field('posData', {'orderId' => order_id}.to_json)
           add_field('fullNotifications', true)
           add_field('transactionSpeed', 'high')
         end
 
+        if options[:checkout_token].length >=44
```
Comment (anbugal) [inline lib/offsite_payments/integrations/bit_pay.rb]: You can reduce this to:

if options[:checkout_token].length >=44 && /^[^0OIl]*$/.match(options[:checkout_token])
    self.invoicing_url = 'https://bitpay.com/invoices'
end

Comment (thejoshualewis) [reply lib/offsite_payments/integrations/bit_pay.rb]: thanks , I updated the PR.

@pi3r  this should be a cleaner implementation and switch the API endpoint based on 
the token length and if certain characters are present that would only exist in the older version

Comment (pi3r) [reply lib/offsite_payments/integrations/bit_pay.rb]: Hey Joshua, thanks for this. Could you add a test case to cover this change? Also this seems to be causing the Bitpay unit tests to fail.

**Comment by thejoshualewis** [comment]
No problem.

I don’t believe we have any test cases for any of the methods, so not really sure how to implement.  This change shouldn’t be breaking anything though, I saw those errors earlier but it looks like its a TravisCI issue.

**Comment by thejoshualewis** [comment]
The errors should be fixed now.

**Comment by pi3r** [comment]
I understand the tests are fixed but we still need a test to prove it's working. You're using `:checkout_token.length` which is always gonna return `14` because `checkout_token` is `14` characters long. Could you please explain exactly the change you're trying to make and give token examples for all scenarios?

**Comment by thejoshualewis** [comment]
@pi3r I'm confused as to why the token would only be 14 characters, maybe the wrong variable is being referenced?

If a user passes in their v1 token, it *should" be <=43 in length.  The new v2 would be >=44

I can email some sample keys privately to show instead of posting them here, would you happen to have an example of one thats only 14 characters that I could then show our developers and verify we're looking at the same thing?

**Comment by pi3r** [comment]
Hey Joshua, what I'm saying is this `:checkout_token` is just a symbol, it does not contain any value.  You can email me some example but also the exact use case you're trying to fix. Like for which call `invoicing_url` needs to be different.

**Pull request closed without merging**